src/unistd.rs

```diff
@@ -2136,3 +2146,424 @@ mod setres {
         Errno::result(res).map(drop)
     }
 }
+
+
+#[derive(Debug, Clone, PartialEq, Default)]
+/// Representation of a user, based on `libc::passwd`
+///
+/// The reason some fields in this struct are `String` and others are `CString` is because some
+/// fields are based on the user's locale, which could be non-UTF8, while other fields are
+/// guaranteed to conform to [NAME_REGEX](https://serverfault.com/a/73101/407341), which only
+/// contains ASCII.
+pub struct User {
+    /// username
+    pub name: String,
+    /// user password (probably encrypted)
+    pub passwd: CString,
+    /// user ID
+    pub uid: Uid,
+    /// group ID
+    pub gid: Gid,
+    #[cfg(not(target_os = "android"))]
+    /// user information
+    pub gecos: CString,
+    /// home directory
+    pub dir: PathBuf,
+    /// path to shell
+    pub shell: PathBuf,
+    #[cfg(not(any(target_os = "linux", target_os = "android")))]
+    /// login class
+    pub class: CString,
+    #[cfg(not(any(target_os = "linux", target_os = "android")))]
+    /// last password change
+    pub change: libc::time_t,
+    #[cfg(not(any(target_os = "linux", target_os = "android")))]
+    /// expiration time of account
+    pub expire: libc::time_t
+}
+
+impl From<*mut libc::passwd> for User {
+    fn from(pw: *mut libc::passwd) -> User {
+        unsafe {
+            User {
+                name: CStr::from_ptr((*pw).pw_name).to_string_lossy().into_owned(),
+                passwd: CString::new(CStr::from_ptr((*pw).pw_passwd).to_bytes()).unwrap(),
+                #[cfg(not(target_os = "android"))]
+                gecos: CString::new(CStr::from_ptr((*pw).pw_gecos).to_bytes()).unwrap(),
+                dir: PathBuf::from(OsStr::from_bytes(CStr::from_ptr((*pw).pw_dir).to_bytes())),
+                shell: PathBuf::from(OsStr::from_bytes(CStr::from_ptr((*pw).pw_shell).to_bytes())),
+                uid: Uid::from_raw((*pw).pw_uid),
+                gid: Gid::from_raw((*pw).pw_gid),
+                #[cfg(not(any(target_os = "linux", target_os = "android")))]
+                class: CString::new(CStr::from_ptr((*pw).pw_class).to_bytes()).unwrap(),
+                #[cfg(not(any(target_os = "linux", target_os = "android")))]
```

Comment:
The OSes need to be alphabetized. Additionally, these sould not use a negation and instead explicitly list all OSes.

Reply:
alphabetical order done; avoiding negations seems quite verbose to me, and `cfg(not(` seems to occur more througout the existing part as well.